applications/smart-distancing/smart_distancing/__main__.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/python3
+import os
+import logging
+
+import smart_distancing as sd
+
+from typing import (
+    Union,
+)
+
+logger = logging.getLogger(__name__)
+
+__all__ = ['main']
+
+def load_config(config: str) -> sd.core.ConfigEngine:
```

Comment:
The reason I did that is in cli_main, I set up handlers on the [root logger](https://docs.python.org/3/library/logging.html#logging.basicConfig), which is the global logging config all instances of Logger use by default. The config file specifies the log path, so it needs to be loaded to set this up. Unless the `force` keyword argument is used (new in 3.8), this setup can only be performed once, which is why it's isolated in cli_main.

I can move it to the distancing module and/or class and set up the handler there, no problem, but if the handlers are set on logger instances rather than globally with logging.basicConfig(), the log settings will just apply to that module/class and have to be set up separately. This is totally doable however if you want differnent log files/formatting/level settings for different classes, for example.

Reply:
Hmm, I know its naive but what about sharing the logger object into the DefaultDistancing module and sub-modules? Seems messy to myself too, but I am thinking of wrapping everything into another custom logger class, both to demonstrate an abstract logger and to keep it simple as passing the instance (I imagine what will we do after e.g. a year, passing everything down? :)) Is there any better architecture at all? .. )

Reply:
So, you can if you want, but i'd recommend against implementing your own logger class since the logging module covers many edge cases (different threads, multiprocessing, stream handling, formatting, etc). So, like, it has locks to ensure messages don't get mangled when two processes attempt to write to the same stream/file, and it handles inter-process communication for you. It also hooks into any loggers that might be in other python modules you include (like PIL, which spammed the log, but will also log to your log file if something goes wrong).

Re: sharing an instance of logger. The way the logging module is implemented, repeated calls to logging.getLogger('name_here') actually return the [same underlying logger](https://docs.python.org/3/library/logging.html#logger-objects). The only difference is the `name_here`, which you can set to anything but generally is set up to be the module or class `__name__`. that way when the logger sends a message to the stream, you have the option to include the origin.

The way it's designed, you generally set the handler (file, stream, syslogd, etc) globally and all instances of logger obtained with logging.getLogger(), no matter the python module, internal or external, hook into it. If you'd like to change it's functionality, my recommendation would be to [subclass something like a Handler or Formatter class](https://github.com/python/cpython/tree/3.8/Lib/logging), allowing you to change the underlying stream handling and formatting, respectively.

Is there something better than the logging module? There might be, but it wouldn't hook into all the other loggers that other modules use. TensorRt has it's own logger, becuase it's not written in Python, but I wrapped that in `...jetson._detectors.py` to hook into the logging module (`LoggingModuleTrtLogger`). If you do choose to implement your own Logger, you can either delete it, or re-implement the log method on that class.